### PR TITLE
Adds channel support with actix-web

### PR DESCRIPTION
**Pull request by eeyun**
Signed-off-by: Ian Henry <ihenry@chef.io>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/builder-api/src/server/resources/channels.rs [hunk 1]
```diff
@@ -31,354 +33,314 @@ impl Channels {
     //
     // Internal - these functions should return Result<..>
     //
+    fn do_promote_package(
+        req: &HttpRequest<AppState>,
+        origin: String,
+        channel: String,
+    ) -> Result<NetOk> {
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: You can just have this be Result<()>

File: components/builder-api/src/server/resources/channels.rs [hunk 2]
```diff
@@ -31,354 +33,314 @@ impl Channels {
...
+        let ident = OriginPackageIdent::new();
+
+        if !helpers::check_origin_access(req, &origin).is_err() {
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: You can change this to
```
helpers::check_origin_access(req, &origin)?;
```

File: components/builder-api/src/server/resources/channels.rs [hunk 3]
```diff
@@ -31,354 +33,314 @@ impl Channels {
...
+        origin_get.set_name(ident.get_origin().to_string());
 
-    //
-    // Route registration
-    //
-    pub fn register(app: App<AppState>) -> App<AppState> {
-        app
-    }
-}
+        let origin_id = match route_message::<OriginGet, Origin>(req, &origin_get) {
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: Mjght be a good time to look whether we need all these Channel.. errors - I think we can just remove channel specific errors as they are not getting used upstream ever... so this paragraph essentially becomes:
```
let origin_id = route_message::<OriginGet, Origin>(req, &origin_get)?;
```

File: components/builder-api/src/server/resources/channels.rs [hunk 4]
```diff
@@ -31,354 +33,314 @@ impl Channels {
...
+            }
+        };
+
+        let mut promote = OriginPackagePromote::new();
+        promote.set_channel_id(origin_channel.get_id());
+        promote.set_package_id(package.get_id());
+        promote.set_ident(ident.clone());
+
+        match route_message::<OriginPackagePromote, NetOk>(req, &promote) {
+            Ok(_) => match route_message::<OriginPackagePromote, NetOk>(req, &promote) {
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: Why are there two OriginPackagePromote calls?

File: components/builder-api/src/server/resources/channels.rs [hunk 5]
```diff
@@ -31,354 +33,314 @@ impl Channels {
...
+                Ok(_) => match Self::audit_package_rank_change(
+                    req,
+                    package.get_id(),
+                    origin_channel.get_id(),
+                    PackageChannelOperation::Promote,
+                    origin_id.get_id(),
+                ) {
+                    Ok(rank) => return Ok(rank),
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: not clear we need to return anything from the audit_package call...

File: components/builder-api/src/server/resources/channels.rs [hunk 6]
```diff
@@ -31,354 +33,314 @@ impl Channels {
...
-                .collect();
-            let body = serde_json::to_string(&list).unwrap();
-            let mut response = Response::with((status::Ok, body));
-            dont_cache_response(&mut response);
-            Ok(response)
+    fn do_demote_package(
+        req: &HttpRequest<AppState>,
+        origin: String,
+        channel: String,
+    ) -> Result<NetOk> {
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: Same, return Result<()>..

File: components/builder-api/src/server/resources/channels.rs [hunk 7]
```diff
@@ -31,354 +33,314 @@ impl Channels {
...
+        route_message::<PackageChannelAudit, NetOk>(req, &audit)
+    }
 
-    let mut request = OriginPackageGet::new();
-    request.set_ident(ident.clone());
-    request.set_visibilities(all_visibilities());
+    //
+    // Route handlers - these functions should return HttpResponse
+    //
+    pub fn list_channels(req: &HttpRequest<AppState>) -> HttpResponse {
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: change "list_channels" to "get_channels" to reflect the more consistent naming strategy called out in the doc..

File: components/builder-api/src/server/resources/channels.rs (hunk 4 above)
Comment (eeyun) [reply components/builder-api/src/server/resources/channels.rs]: I'm not sure on that, this is how it is in master right now.

File: components/builder-api/src/server/resources/channels.rs [hunk 8]
```diff
@@ -31,354 +33,314 @@ impl Channels {
...
-        Some(channel) => channel,
-        None => return Ok(Response::with(status::BadRequest)),
-    };
-
-    // you can't demote from "unstable"
-    if channel == "unstable" {
-        return Ok(Response::with(status::Forbidden));
+        // Pass ?sandbox=true to this endpoint to include sandbox channels in the list. They are not
+        // there by default.
+        let sandbox = Query::<(String)>::extract(&req).unwrap().into_inner();
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: You can put this query param handling in the function prototype, and have it automatically parse into a struct... see the code sample from one of the other modules (eg pkgs.rs).. lot cleaner

File: components/builder-api/src/server/resources/channels.rs (hunk 4 above)
Comment (chefsalim) [reply components/builder-api/src/server/resources/channels.rs]: I think it's a copy-pasta error - we can just call promote once..

File: components/builder-api/src/server/resources/channels.rs [hunk 9]
```diff
@@ -31,354 +33,314 @@ impl Channels {
...
+    }
 
-    let mut request = OriginPackageGet::new();
-    request.set_ident(ident.clone());
-    request.set_visibilities(all_visibilities());
+    //
+    // Route handlers - these functions should return HttpResponse
+    //
+    pub fn list_channels(req: &HttpRequest<AppState>) -> HttpResponse {
+        let origin = Path::<(String)>::extract(&req).unwrap().into_inner(); //Unwrap apparently Ok
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: The unwrap here is definitely ok.. the handler will never get called if there is no string to match in the route

File: components/builder-api/src/server/resources/channels.rs [hunk 10]
```diff
@@ -14,371 +14,326 @@
...
+
+        let mut channel_req = OriginChannelGet::new();
+        channel_req.set_origin_name(ident.get_origin().to_string());
+        channel_req.set_name(channel.to_string());
+
+        let origin_channel =
+            match route_message::<OriginChannelGet, OriginPackage>(req, &channel_req) {
+                Ok(c) => c,
+                Err(e) => {
+                    error!("Error retrieving channel: {:?}", &e);
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: minor - switch error! to warn! .. maybe even debug! if these can happen under 'normal' operation

File: components/builder-api/src/server/resources/channels.rs [hunk 11]
```diff
@@ -14,371 +14,326 @@
...
 
-    let package = match route_message::<OriginPackageGet, OriginPackage>(req, &request) {
-        Ok(p) => p,
-        Err(e) => {
-            error!("Error retrieving package: {:?}", &e);
-            return Ok(render_net_error(&e));
-        }
-    };
+        let mut request = OriginChannelListRequest::new();
+        request.set_include_sandbox_channels(false);
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: use sandbox.is_set in the param?

File: components/builder-api/src/server/resources/channels.rs [hunk 12]
```diff
@@ -14,371 +14,326 @@
...
+                    .map(|channel| {
+                        let mut ident = OriginChannelIdent::new();
+                        ident.set_name(channel.get_name().to_string());
+                        ident
+                    })
+                    .collect();
+                let body = serde_json::to_string(&list).unwrap();
+                let mut response = HttpResponse::Ok();
+                response
+                    .header(http::header::CACHE_CONTROL, "private, no-cache, no-store")
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: use const from middleware/headers.rs for the cache control string..

File: components/builder-api/src/server/resources/channels.rs [hunk 13]
```diff
@@ -14,371 +14,326 @@
...
+            Err(err) => err.into(),
+        }
+    }
 
-    route_message::<PackageChannelAudit, NetOk>(req, &audit)
+    //
+    // Route registration
+    //
+    pub fn register(app: App<AppState>) -> App<AppState> {
+        app.resource("/channels/:origin", |r| {
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/channels.rs]: change the route registration params .. eg from ":origin" to "{origin}" (for example)
I missed that when wiring up some of the modules :)

**Pull request merged**